### DISCUSSION TITLE
Proper way to add an item to StandardListView?

### DISCUSSION THREAD

**Discussion by AllenDang** [slint-ui/slint#2329] [Q&A] [answered]
My current implementation kind of works, but not ideal, here is the code.

```rust
let mut logs: Vec<StandardListViewItem> = ui.get_log_model().iter().collect();
logs.push(StandardListViewItem {
    text: some_text.clone().into(),
});

let logs_model = Rc::new(slint::VecModel::from(logs));
ui.set_log_model(logs_model.into());
```

```
    in property <[StandardListViewItem]> log_model;
    ...
        StandardListView { 
          height: 200px;
          model: root.log_model;
          viewport-y: min(0px, self.visible-height - self.viewport-height);
        }

```

It's kind of wired, get whole vector of StandardListViewItem, push new, set the vector to the model, maybe I'm not doing it correct?

Ideally I should do something like `ui.get_log_model().push(...)`.

**Comment by bjorn** [comment]
What does your `ui.get_log_model()` return? Did you try using [VecModel::push](https://slint-ui.com/releases/0.3.5/docs/rust/slint/struct.vecmodel#method.push)?

**Reply by AllenDang** [reply]
![image](https://user-images.githubusercontent.com/18550/223084191-6a2c98a7-d6af-4cfd-b42c-9bc392b353e5.png)

**Comment by ogoffart** [comment, marked as answer]
The idea is that you would keep a reference to the `VecModel` and you can add your log like that.

You can also do `ui.get_log_model().as_any().downcast_ref::<slint::VecModel<StandardListViewItem>>()`  to get the VecModel back (assuming you initialized the model with a VecModel before)

**Reply by AllenDang** [reply]
I thought the initialization is done by slint internally, should I manually set that from rust side?

**Reply by ogoffart** [reply]
> I thought the initialization is done by slint internally, should I manually set that from rust side?

Yes, slint could also create something else than a VecModel

**Reply by AllenDang** [reply]
I see, the conversion works if I manually initiate the model, thx!

**Reply by AllenDang** [reply]
Is that a potential bug for slint? I define the property as `property <[StandardListItem>]` but it doesn't initiate as that?

**Reply by ogoffart** [reply]
Well, Slint could use any kind of model which might not be a VecModel.  For example, it could use a model to a static array, or an empty model.
